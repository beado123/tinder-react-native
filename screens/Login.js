import React from 'react';
import styles from '../styles/Login-style.js';
import RootNavigator from '../navigation/RootNavigator.js';
import  {connect} from 'react-redux';
import {login, updateNameGender} from '../redux/actions.js';
import { Text, View, ScrollView, Alert, TouchableOpacity, ImageBackground } from 'react-native';
import * as firebase from 'firebase';
import firebaseConfig from '../config/firebase.js';
import { YellowBox } from 'react-native';
firebase.initializeApp(firebaseConfig);

export class Login extends React.Component {

    state = {
        name: this.props.user.name==''?'':this.props.user.name,
        birthday: this.props.user.birth==''?'':this.props.user.birth,
        avatar_url: this.props.user.photoUrl==''?'':this.props.user.photoUrl,
    }

    /*Attach an observer of user-login state in firebase*/
    componentDidMount(){
        firebase.auth().onAuthStateChanged((user) => {
            //get current signed in User
            console.log('firebase onAuthStateChanged..');
            if(user != null){
                console.log("Authenticating: " + JSON.stringify(user));
                this.props.dispatch(login(user,this.state.name,this.state.birthday,this.state.avatar_url));
            }
        });
    }

    /*Enable Facebook login and authenticate firebase. */
    login = async () => {

        const { type, token } = await Expo.Facebook.logInWithReadPermissionsAsync('211190339637510', {
            permissions: ['public_profile', 'user_birthday'],
        });
        if (type === 'success') {
            // build firebase credentials
            const credential = await firebase.auth.FacebookAuthProvider.credential(token);
            fetch('https://graph.facebook.com/me?fields=birthday,name,id&access_token='+token)
            .then(response => response.json())
            .then(function(responseJson){
                console.log('object: '+JSON.stringify(responseJson));
                this.setState({birthday: responseJson.birthday, name: responseJson.name});
                //get avatar from facebook
                fetch('https://graph.facebook.com/'+responseJson.id+'/picture?type=square')
                    .then(function(avatarJson){
                        console.log('getting avatar, '+JSON.stringify(avatarJson));
                        this.setState({avatar_url: avatarJson.url});
                        //sign in with credential
                        firebase.auth().signInWithCredential(credential).catch((error) => {
                            Alert.alert("Try Again")
                        });
                    }.bind(this)).catch((error) => {
                        console.error(error);
                    });
            }.bind(this)).catch((error) => {
                console.error(error);
            });
        }
    }

    render() {
        if(this.props.loggedIn){
            return(
                <RootNavigator/>
            )
        }else{
            return (
                <ScrollView contentContainerStyle={{flexGrow: 1,}} scrollEnabled>
                <View style={styles.container}>
                    <ImageBackground style={styles.headerBackground} source={require('/home/yihan/Tinder/assets/tinder_background.png')}>
                        <View style={{flex:3}}>
                            <Text style={styles.logo}>Tinder</Text>
                        </View>
                        <View style={styles.content}>
                            <TouchableOpacity style = {styles.buttonLogin} onPress = {this.login.bind(this)}>
                                <Text style={styles.buttonText}>LOGIN</Text>
                            </TouchableOpacity>
                        </View>
                    </ImageBackground>
                    </View>
                </ScrollView>
            );
        }
    }
}

/*Connect state from the store to props, state.loggedIn corresponds to action.loggedIn in reducer.js*/
function mapStateToProps(state) {
    console.log('loading login page');
    return {
        loggedIn: state.loggedIn,
        user: state.user
    };
}

/*Connects a React component to a Redux store*/
export default connect(mapStateToProps)(Login);
