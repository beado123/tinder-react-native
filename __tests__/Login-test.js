import React from 'react';
import {Login} from '../screens/Login.js';
import renderer from 'react-test-renderer';

test('renders correctly', () => {
  const tree = renderer.create(<Login user={{photoUrl: 'https//mock_email', name: 'mock_name'}}  />).toJSON();
  expect(tree).toMatchSnapshot();
});
