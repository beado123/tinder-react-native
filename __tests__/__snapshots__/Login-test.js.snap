// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  scrollEnabled={true}
>
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "flex": 1,
          }
        }
      >
        <Image
          source={1}
          style={
            Array [
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              Object {
                "height": undefined,
                "width": undefined,
              },
              undefined,
            ]
          }
        />
        <View
          style={
            Object {
              "flex": 3,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#fff",
                "fontFamily": "Georgia",
                "fontSize": 40,
                "marginTop": 80,
                "textAlign": "center",
                "textShadowColor": "#ffe1ff",
                "textShadowOffset": Object {
                  "height": 2,
                  "width": 2,
                },
                "textShadowRadius": 15,
                "width": 400,
              }
            }
          >
            Tinder
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "column",
            }
          }
        >
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            collapsable={undefined}
            hasTVPreferredFocus={undefined}
            hitSlop={undefined}
            isTVSelectable={true}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "#ff6a80",
                "borderRadius": 20,
                "opacity": 1,
                "padding": 15,
              }
            }
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#fff",
                  "fontFamily": "Georgia",
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "textAlign": "center",
                }
              }
            >
              LOGIN
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
